addons/controllers/cpi/vspherecpiconfig_utils.go

```diff
@@ -20,13 +20,62 @@ import (
 	capvv1beta1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/v1beta1"
 	capvvmwarev1beta1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/vmware/v1beta1"
 	clusterapiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	clusterapiutil "sigs.k8s.io/cluster-api/util"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/constants"
 	pkgtypes "github.com/vmware-tanzu/tanzu-framework/addons/pkg/types"
 	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/util"
 	cpiv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/addonconfigs/cpi/v1alpha1"
 )
 
+// VSphereClusterToVSphereCPIConfig returns a list of Requests with VSphereCPIConfig ObjectKey based on Cluster events
+func (r *VSphereCPIConfigReconciler) VSphereClusterToVSphereCPIConfig(o client.Object) []ctrl.Request {
+	cluster, ok := o.(*capvvmwarev1beta1.VSphereCluster)
+	if !ok {
+		r.Log.Error(errors.New("invalid type"),
+			"Expected to receive Cluster resource",
+			"actualType", fmt.Sprintf("%T", o))
+		return nil
+	}
+
+	r.Log.V(4).Info("Mapping VSphereCluster to VSphereCPIConfig")
+
+	cs := &cpiv1alpha1.VSphereCPIConfigList{}
+	_ = r.List(context.Background(), cs)
+
+	requests := []ctrl.Request{}
+	for i := 0; i < len(cs.Items); i++ {
+		config := &cs.Items[i]
+		if config.Namespace == cluster.Namespace {
+
+			// avoid enqueuing reconcile requests for template vSphereCPIConfig CRs in event handler of Cluster CR
+			if _, ok := config.Annotations[constants.TKGAnnotationTemplateConfig]; ok && config.Namespace == r.Config.SystemNamespace {
+				continue
+			}
+
+			// corresponding vsphereCPIConfig should have following ownerRef
+			ownerReference := metav1.OwnerReference{
+				APIVersion: clusterapiv1beta1.GroupVersion.String(),
+				Kind:       cluster.Kind,
+				Name:       cluster.Name,
+				UID:        cluster.UID,
+			}
+			if clusterapiutil.HasOwnerRef(config.OwnerReferences, ownerReference) || config.Name == fmt.Sprintf("%s-%s-package", cluster.Name, constants.CPIAddonName) {
```

Comment:
I am confused with `fmt.Sprintf("%s-%s-package", cluster.Name, constants.CPIAddonName)` as package name and cluster name as package name.

Here we are filtering the cpiConfig to enqueue those who has ownRef to an existing clusters, or who has the <clustername>-<package-prefix>-package name where the cluster is exist.

But in here we only reconcile those who has ownRef to an existing clusters, our who has the name of an existing cluster.

https://github.com/lubronzhan/tanzu-framework/blob/975271dee7d6ecb0c60fc271d1d29805ecf6f18c/addons/controllers/cpi/vspherecpiconfig_utils.go#L403-L427

Reply:
And ALL the AddonConfig test cases are based on `cluster name as cpiConfig name` for controller testing. Like

https://github.com/lubronzhan/tanzu-framework/blob/975271dee7d6ecb0c60fc271d1d29805ecf6f18c/addons/controllers/testdata/test-vsphere-cpi-non-paravirtual.yaml

This means **the user customized AddonConfig which has name `"clustername"-"package-prefix"-package` will never be reconciled, and which has name of the cluster name will never be enqueued base on the cluster event.**

The confusion between `cluster name as AddonConfig name` and `"cluster name"-"package prefix"-package` as AddonConfig name has bring a lot of unsync, I think we really should make this clear.

Per my understanding, the name format of `"cluster name"-"package prefix"-package` was never a part of the design doc, yet originally generated based on the copy of the ClusterBootstrapTemplate. All other customized cases are following the cluster name as AddonConfig name. Now this naming conversion becomes more and more conflict. I think we should eliminate the concept of one of them totally.

cc @vijaykatam @DanielXiao @HanFa

Reply:
We are relying on "cluster name"-"package shortName"-package now. The original design was for "cluster name" but a refactor(I think cloning for webhook) changed it. I would like us to continue down this path of updated name because this has already shipped with vSphere 8.0 integration.